In [102]:
#Antes de nada voy a importar las librerías necesarias 
import pandas as pd 
import numpy as np 
import seaborn as sns
import matplotlib.pyplot as plt

In [103]:
#En segundo lugar, lo configuro para poder visualizar todas las columnas de los DataFrames
pd.set_option("display.max_columns", None)

Voy a comenzar leyendo cada uno de los csv

In [ ]:
#Leo el csv de datos de 2013 y observo las primeras filas de este csv
df_2013 = pd.read_csv("datos/datos-2013.csv", sep=";", encoding="ISO-8859-1")
df_2013.head()

CÓDIGO ÓRGÃO SUPERIOR       NOME ÓRGÃO SUPERIOR  CÓDIGO ÓRGÃO  \
0                  63000  Advocacia-Geral da União         63000   
1                  63000  Advocacia-Geral da União         63000   
2                  63000  Advocacia-Geral da União         63000   
3                  63000  Advocacia-Geral da União         63000   
4                  63000  Advocacia-Geral da União         63000   

                                          NOME ÓRGÃO  CÓDIGO UNIDADE GESTORA  \
0  Advocacia-Geral da União - Unidades com víncul...                  110060   
1  Advocacia-Geral da União - Unidades com víncul...                  110060   
2  Advocacia-Geral da União - Unidades com víncul...                  110060   
3  Advocacia-Geral da União - Unidades com víncul...                  110060   
4  Advocacia-Geral da União - Unidades com víncul...                  110060   

                            NOME UNIDADE GESTORA CATEGORIA ECONÔMICA  \
0  COORD. GERAL DE ORC. FIN. E ANAL. CONT. - AGU  Receitas Correntes   
1  COORD. GERAL DE ORC. FIN. E ANAL. CONT. - AGU  Receitas Correntes   
2  COORD. GERAL DE ORC. FIN. E ANAL. CONT. - AGU  Receitas Correntes   
3  COORD. GERAL DE ORC. FIN. E ANAL. CONT. - AGU  Receitas Correntes   
4  COORD. GERAL DE ORC. FIN. E ANAL. CONT. - AGU  Receitas Correntes   

              ORIGEM RECEITA                                ESPÉCIE RECEITA  \
0  Outras Receitas Correntes  Bens, Direitos e Valores Incorporados ao Patr   
1  Outras Receitas Correntes    Indenizações, restituições e ressarcimentos   
2  Outras Receitas Correntes  Multas administrativas, contratuais e judicia   
3  Outras Receitas Correntes  Bens, Direitos e Valores Incorporados ao Patr   
4  Outras Receitas Correntes    Indenizações, restituições e ressarcimentos   

                                    DETALHAMENTO VALOR PREVISTO ATUALIZADO  \
0  REC.DIVIDA ATIVA NAO TRIBUTARIA DE OUTRAS REC                      0,00   
1   RECUPERACAO DE DESPESAS DE EXERC. ANTERIORES                      0,00   
2                  OUTRAS MULTAS E JUROS DE MORA                      0,00   
3      REC.DIV.ATIVA POR INFRAÇÃO ADMINISTRATIVA                      0,00   
4                            OUTRAS RESTITUICOES                      0,00   

  VALOR LANÇADO VALOR REALIZADO PERCENTUAL REALIZADO DATA LANÇAMENTO  \
0          0,00         1297,13                 0,00      31/12/2013   
1          0,00     26666621,42                 0,00      31/12/2013   
2          0,00       301251,13                 0,00      31/12/2013   
3          0,00         1855,58                 0,00      31/12/2013   
4          0,00        52140,68                 0,00      31/12/2013   

   ANO EXERCÍCIO  
0           2013  
1           2013  
2           2013  
3           2013  
4           2013

In [105]:
#Me saco el numero de filas y columnas de este dataframe
df_2013.shape

(4498, 16)

Para el año 2013 contamos con 16 columnas y 4498 filas

In [106]:
#Me saco el nombre de las columnas porque voy a renombrarlas para obtenerlas en español 
"""df_2013.columns"""

'df_2013.columns'

In [107]:
#Diccionario que me ha pasado chatgpt con las indicaciones de traduccion que le he metido para hacer la traduccion de las columnas 
"""traduccion_columnas = {
    "CÓDIGO ÓRGÃO SUPERIOR": "Código_Órgano_Superior",
    "NOME ÓRGÃO SUPERIOR": "Nombre_Órgano_Superior",
    "CÓDIGO ÓRGÃO": "Código_Órgano",
    "NOME ÓRGÃO": "Nombre_Órgano",
    "CÓDIGO UNIDADE GESTORA": "Código_Unidad_Gestora",
    "NOME UNIDADE GESTORA": "Nombre_Unidad_Gestora",
    "CATEGORIA ECONÔMICA": "Categoría_Económica",
    "ORIGEM RECEITA": "Origen_Ingreso",
    "ESPÉCIE RECEITA": "Tipo_Ingreso",
    "DETALHAMENTO": "Detalle",
    "VALOR PREVISTO ATUALIZADO": "Valor_Previsto_Actualizado",
    "VALOR LANÇADO": "Valor_Registrado",
    "VALOR REALIZADO": "Valor_Realizado",
    "PERCENTUAL REALIZADO": "Porcentaje_Realizado",
    "DATA LANÇAMENTO": "Fecha_Registro",
    "ANO EXERCÍCIO": "Año_Ejercicio"
}"""


'traduccion_columnas = {\n    "CÓDIGO ÓRGÃO SUPERIOR": "Código_Órgano_Superior",\n    "NOME ÓRGÃO SUPERIOR": "Nombre_Órgano_Superior",\n    "CÓDIGO ÓRGÃO": "Código_Órgano",\n    "NOME ÓRGÃO": "Nombre_Órgano",\n    "CÓDIGO UNIDADE GESTORA": "Código_Unidad_Gestora",\n    "NOME UNIDADE GESTORA": "Nombre_Unidad_Gestora",\n    "CATEGORIA ECONÔMICA": "Categoría_Económica",\n    "ORIGEM RECEITA": "Origen_Ingreso",\n    "ESPÉCIE RECEITA": "Tipo_Ingreso",\n    "DETALHAMENTO": "Detalle",\n    "VALOR PREVISTO ATUALIZADO": "Valor_Previsto_Actualizado",\n    "VALOR LANÇADO": "Valor_Registrado",\n    "VALOR REALIZADO": "Valor_Realizado",\n    "PERCENTUAL REALIZADO": "Porcentaje_Realizado",\n    "DATA LANÇAMENTO": "Fecha_Registro",\n    "ANO EXERCÍCIO": "Año_Ejercicio"\n}'

In [108]:
#renombro las columnas
"""df_2013.rename(columns=traduccion_columnas, inplace=True)
df_2013.head()"""

'df_2013.rename(columns=traduccion_columnas, inplace=True)\ndf_2013.head()'

In [117]:
#Leo el csv de datos de 2014 
df_2014 = pd.read_csv("datos/datos-2014.csv", sep=";", encoding="ISO-8859-1")
df_2014.head()

CÓDIGO ÓRGÃO SUPERIOR       NOME ÓRGÃO SUPERIOR  CÓDIGO ÓRGÃO  \
0                  63000  Advocacia-Geral da União         63000   
1                  63000  Advocacia-Geral da União         63000   
2                  63000  Advocacia-Geral da União         63000   
3                  63000  Advocacia-Geral da União         63000   
4                  63000  Advocacia-Geral da União         63000   

                                          NOME ÓRGÃO  CÓDIGO UNIDADE GESTORA  \
0  Advocacia-Geral da União - Unidades com víncul...                  110060   
1  Advocacia-Geral da União - Unidades com víncul...                  110060   
2  Advocacia-Geral da União - Unidades com víncul...                  110060   
3  Advocacia-Geral da União - Unidades com víncul...                  110060   
4  Advocacia-Geral da União - Unidades com víncul...                  110060   

                            NOME UNIDADE GESTORA  \
0  COORD. GERAL DE ORC. FIN. E ANAL. CONT. - AGU   
1  COORD. GERAL DE ORC. FIN. E ANAL. CONT. - AGU   
2  COORD. GERAL DE ORC. FIN. E ANAL. CONT. - AGU   
3  COORD. GERAL DE ORC. FIN. E ANAL. CONT. - AGU   
4  COORD. GERAL DE ORC. FIN. E ANAL. CONT. - AGU   

                        CATEGORIA ECONÔMICA             ORIGEM RECEITA  \
0                        Receitas Correntes  Outras Receitas Correntes   
1                        Receitas Correntes  Outras Receitas Correntes   
2                        Receitas Correntes  Outras Receitas Correntes   
3  Receitas Correntes - intra-orçamentárias  Outras Receitas Correntes   
4                        Receitas Correntes        Receita de Serviços   

                                 ESPÉCIE RECEITA  \
0                      Demais receitas correntes   
1  Multas administrativas, contratuais e judicia   
2  Multas administrativas, contratuais e judicia   
3                    Indenizações e restituições   
4                            Receita de Serviços   

                                   DETALHAMENTO VALOR PREVISTO ATUALIZADO  \
0            RECEITA DE HONORÁRIOS DE ADVOGADOS                      0,00   
1         MULTAS E JUROS PREVISTOS EM CONTRATOS                      0,00   
2                 OUTRAS MULTAS E JUROS DE MORA                      0,00   
3  RECUPERACAO DE DESPESAS DE EXERC. ANTERIORES                      0,00   
4     TAR.INSCR.CONCURSOS E PROCESSOS SELETIVOS                      0,00   

  VALOR LANÇADO VALOR REALIZADO PERCENTUAL REALIZADO DATA LANÇAMENTO  \
0          0,00     46699948,08                 0,00      31/12/2014   
1          0,00        15920,83                 0,00      31/12/2014   
2          0,00      2195899,92                 0,00      31/12/2014   
3          0,00          232,02                 0,00      31/12/2014   
4          0,00       367815,01                 0,00      31/12/2014   

   ANO EXERCÍCIO  
0           2014  
1           2014  
2           2014  
3           2014  
4           2014

In [118]:
#Me saco el numero de filas y columnas de este dataframe
df_2014.shape

(4553, 16)

Para 2014 contamos con 16 columnas y 4553 filas. Vamos a comprobar que las 16 columnas son las mismas

In [119]:
df_2013.sample()

CÓDIGO ÓRGÃO SUPERIOR     NOME ÓRGÃO SUPERIOR  CÓDIGO ÓRGÃO  \
2471                  26000  Ministério da Educação         26285   

                                             NOME ÓRGÃO  \
2471  Fundação Universidade Federal de São João Del-Rei   

      CÓDIGO UNIDADE GESTORA                          NOME UNIDADE GESTORA  \
2471                  154069  FUNDACAO UNIVERSIDADE FEDERAL DE S.J.DEL-REI   

     CATEGORIA ECONÔMICA       ORIGEM RECEITA  \
2471  Receitas Correntes  Receita Patrimonial   

                                    ESPÉCIE RECEITA   DETALHAMENTO  \
2471  Exploração do patrimônio imobiliário do Estad  ARRENDAMENTOS   

     VALOR PREVISTO ATUALIZADO VALOR LANÇADO VALOR REALIZADO  \
2471                 144176,00          0,00       140297,20   

     PERCENTUAL REALIZADO DATA LANÇAMENTO  ANO EXERCÍCIO  
2471                97,00      31/12/2013           2013

In [120]:
df_2014.sample()

CÓDIGO ÓRGÃO SUPERIOR     NOME ÓRGÃO SUPERIOR  CÓDIGO ÓRGÃO  \
2272                  26000  Ministério da Educação         26350   

                                            NOME ÓRGÃO  \
2272  Fundação Universidade Federal da Grande Dourados   

      CÓDIGO UNIDADE GESTORA                           NOME UNIDADE GESTORA  \
2272                  154502  FUNDACAO UNIVERSIDADE FED. DA GRANDE DOURADOS   

     CATEGORIA ECONÔMICA       ORIGEM RECEITA      ESPÉCIE RECEITA  \
2272  Receitas Correntes  Receita de Serviços  Receita de Serviços   

                         DETALHAMENTO VALOR PREVISTO ATUALIZADO VALOR LANÇADO  \
2272  SERVICOS DE ESTUDOS E PESQUISAS                      0,00          0,00   

     VALOR REALIZADO PERCENTUAL REALIZADO DATA LANÇAMENTO  ANO EXERCÍCIO  
2272       116728,54                 0,00      31/12/2014           2014

Efectivamente las columnas son las mismas y en el mismo orden. Revisaré que para todos los años sea así y de esta forma se podrá concatenar perfectamente los diferentes dataframes para obtener el final

Voy a empezar haciendo una pequeña exploración de nulos y de duplicados en cada uno de los dataframes

Para 2013

In [121]:
#Para observar nulos
df_2013.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4498 entries, 0 to 4497
Data columns (total 16 columns):
 #   Column                     Non-Null Count  Dtype 
---  ------                     --------------  ----- 
 0   CÓDIGO ÓRGÃO SUPERIOR      4498 non-null   int64 
 1   NOME ÓRGÃO SUPERIOR        4498 non-null   object
 2   CÓDIGO ÓRGÃO               4498 non-null   int64 
 3   NOME ÓRGÃO                 4498 non-null   object
 4   CÓDIGO UNIDADE GESTORA     4498 non-null   int64 
 5   NOME UNIDADE GESTORA       4498 non-null   object
 6   CATEGORIA ECONÔMICA        4498 non-null   object
 7   ORIGEM RECEITA             4498 non-null   object
 8   ESPÉCIE RECEITA            4498 non-null   object
 9   DETALHAMENTO               4498 non-null   object
 10  VALOR PREVISTO ATUALIZADO  4498 non-null   object
 11  VALOR LANÇADO              4498 non-null   object
 12  VALOR REALIZADO            4498 non-null   object
 13  PERCENTUAL REALIZADO       4498 non-null   object
 14  DATA LAN

In [124]:
#Comprobar si hay duplicados por todas las columnas
df_2013.duplicated().sum()

np.int64(0)

In [126]:
#Comprobar si hay duplicados por la columna DETALHAMENTO
df_2013.duplicated("DETALHAMENTO").sum()

np.int64(3794)

Parece que no hay nulos y que tampoco hay duplicados por todas las columnas. En este caso no contamos con una columna que me identifique cada registro, cada ingreso, por lo tanto es complicado observar si hay duplicados por una columna específica. He pensado que quizás la columna de DETALHAMENTO que te da el detalle del ingreso podría ser como el "identificador unico" en esta tabla pero veo que hay un montón de duplicados también por esta columna asique no es posible.

Para 2014

In [130]:
#Para observar nulos
df_2014.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4553 entries, 0 to 4552
Data columns (total 16 columns):
 #   Column                     Non-Null Count  Dtype 
---  ------                     --------------  ----- 
 0   CÓDIGO ÓRGÃO SUPERIOR      4553 non-null   int64 
 1   NOME ÓRGÃO SUPERIOR        4553 non-null   object
 2   CÓDIGO ÓRGÃO               4553 non-null   int64 
 3   NOME ÓRGÃO                 4553 non-null   object
 4   CÓDIGO UNIDADE GESTORA     4553 non-null   int64 
 5   NOME UNIDADE GESTORA       4553 non-null   object
 6   CATEGORIA ECONÔMICA        4553 non-null   object
 7   ORIGEM RECEITA             4553 non-null   object
 8   ESPÉCIE RECEITA            4553 non-null   object
 9   DETALHAMENTO               4553 non-null   object
 10  VALOR PREVISTO ATUALIZADO  4553 non-null   object
 11  VALOR LANÇADO              4553 non-null   object
 12  VALOR REALIZADO            4553 non-null   object
 13  PERCENTUAL REALIZADO       4553 non-null   object
 14  DATA LAN

In [132]:
#Comprobar si hay duplicados por todas las columnas
df_2014.duplicated().sum()

np.int64(0)

Para 2014 parece que tampoco hay nulos ni duplicados

Para 2015

In [133]:
#Leo el csv de datos de 2015
df_2015 = pd.read_csv("datos/datos-2015.csv", sep=";", encoding="ISO-8859-1")
df_2015.head()

CÓDIGO ÓRGÃO SUPERIOR       NOME ÓRGÃO SUPERIOR  CÓDIGO ÓRGÃO  \
0                  63000  Advocacia-Geral da União         63000   
1                  63000  Advocacia-Geral da União         63000   
2                  63000  Advocacia-Geral da União         63000   
3                  63000  Advocacia-Geral da União         63000   
4                  63000  Advocacia-Geral da União         63000   

                                          NOME ÓRGÃO  CÓDIGO UNIDADE GESTORA  \
0  Advocacia-Geral da União - Unidades com víncul...                  110060   
1  Advocacia-Geral da União - Unidades com víncul...                  110060   
2  Advocacia-Geral da União - Unidades com víncul...                  110060   
3  Advocacia-Geral da União - Unidades com víncul...                  110060   
4  Advocacia-Geral da União - Unidades com víncul...                  110060   

                            NOME UNIDADE GESTORA CATEGORIA ECONÔMICA  \
0  COORD. GERAL DE ORC. FIN. E ANAL. CONT. - AGU  Receitas Correntes   
1  COORD. GERAL DE ORC. FIN. E ANAL. CONT. - AGU  Receitas Correntes   
2  COORD. GERAL DE ORC. FIN. E ANAL. CONT. - AGU  Receitas Correntes   
3  COORD. GERAL DE ORC. FIN. E ANAL. CONT. - AGU  Receitas Correntes   
4  COORD. GERAL DE ORC. FIN. E ANAL. CONT. - AGU  Receitas Correntes   

              ORIGEM RECEITA                                ESPÉCIE RECEITA  \
0  Outras Receitas Correntes                      Demais receitas correntes   
1        Receita de Serviços                            Receita de Serviços   
2  Outras Receitas Correntes                      Demais receitas correntes   
3  Outras Receitas Correntes    Indenizações, restituições e ressarcimentos   
4  Outras Receitas Correntes  Multas administrativas, contratuais e judicia   

                                DETALHAMENTO VALOR PREVISTO ATUALIZADO  \
0         RECEITA DE HONORARIOS DE ADVOGADOS                      0,00   
1  TAR.INSCR.CONCURSOS E PROCESSOS SELETIVOS                 636000,00   
2             RECEITA DE ONUS DE SUCUMBENCIA                      0,00   
3                        OUTRAS RESTITUICOES                      0,00   
4              OUTRAS MULTAS E JUROS DE MORA                      0,00   

  VALOR LANÇADO VALOR REALIZADO PERCENTUAL REALIZADO DATA LANÇAMENTO  \
0          0,00     63829853,40                 0,00      31/12/2015   
1          0,00      4486964,41               705,00      31/12/2015   
2          0,00     13060401,15                 0,00      31/12/2015   
3          0,00        48085,58                 0,00      31/12/2015   
4          0,00      1022856,02                 0,00      31/12/2015   

   ANO EXERCÍCIO  
0           2015  
1           2015  
2           2015  
3           2015  
4           2015

In [137]:
#Me saco el numero de filas y columnas de este dataframe
df_2015.shape

(4523, 16)

Para 2015 cuento con 4253 filas y 16 columnas

In [138]:
#Para observar nulos
df_2015.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4523 entries, 0 to 4522
Data columns (total 16 columns):
 #   Column                     Non-Null Count  Dtype 
---  ------                     --------------  ----- 
 0   CÓDIGO ÓRGÃO SUPERIOR      4523 non-null   int64 
 1   NOME ÓRGÃO SUPERIOR        4523 non-null   object
 2   CÓDIGO ÓRGÃO               4523 non-null   int64 
 3   NOME ÓRGÃO                 4523 non-null   object
 4   CÓDIGO UNIDADE GESTORA     4523 non-null   int64 
 5   NOME UNIDADE GESTORA       4523 non-null   object
 6   CATEGORIA ECONÔMICA        4523 non-null   object
 7   ORIGEM RECEITA             4523 non-null   object
 8   ESPÉCIE RECEITA            4523 non-null   object
 9   DETALHAMENTO               4523 non-null   object
 10  VALOR PREVISTO ATUALIZADO  4523 non-null   object
 11  VALOR LANÇADO              4523 non-null   object
 12  VALOR REALIZADO            4523 non-null   object
 13  PERCENTUAL REALIZADO       4523 non-null   object
 14  DATA LAN

In [139]:
#Comprobar si hay duplicados por todas las columnas
df_2015.duplicated().sum()

np.int64(0)

Para 2015 parece que no hay nulos ni duplicados

Para 2016

In [140]:
#Leo el csv de datos de 2016
df_2016 = pd.read_csv("datos/datos-2016.csv", sep=";", encoding="ISO-8859-1")
df_2016.head()

CÓDIGO ÓRGÃO SUPERIOR       NOME ÓRGÃO SUPERIOR  CÓDIGO ÓRGÃO  \
0                  63000  Advocacia-Geral da União         63000   
1                  63000  Advocacia-Geral da União         63000   
2                  63000  Advocacia-Geral da União         63000   
3                  63000  Advocacia-Geral da União         63000   
4                  63000  Advocacia-Geral da União         63000   

                                          NOME ÓRGÃO  CÓDIGO UNIDADE GESTORA  \
0  Advocacia-Geral da União - Unidades com víncul...                  110060   
1  Advocacia-Geral da União - Unidades com víncul...                  110060   
2  Advocacia-Geral da União - Unidades com víncul...                  110060   
3  Advocacia-Geral da União - Unidades com víncul...                  110060   
4  Advocacia-Geral da União - Unidades com víncul...                  110060   

                            NOME UNIDADE GESTORA CATEGORIA ECONÔMICA  \
0  COORD. GERAL DE ORC. FIN. E ANAL. CONT. - AGU  Receitas Correntes   
1  COORD. GERAL DE ORC. FIN. E ANAL. CONT. - AGU  Receitas Correntes   
2  COORD. GERAL DE ORC. FIN. E ANAL. CONT. - AGU  Receitas Correntes   
3  COORD. GERAL DE ORC. FIN. E ANAL. CONT. - AGU  Receitas Correntes   
4  COORD. GERAL DE ORC. FIN. E ANAL. CONT. - AGU  Receitas Correntes   

                                ORIGEM RECEITA            ESPÉCIE RECEITA  \
0                    Outras Receitas Correntes  Demais receitas correntes   
1                    Outras Receitas Correntes  Demais receitas correntes   
2                    Outras Receitas Correntes  Demais receitas correntes   
3  Impostos, Taxas e Contribuições de Melhoria                      Taxas   
4                    Outras Receitas Correntes  Demais receitas correntes   

                                    DETALHAMENTO VALOR PREVISTO ATUALIZADO  \
0  ENCARGOS LEGAIS PELA INSCR.EM DIV.ATIVA-PRINC                      0,00   
1             ONUS DE SUCUMBENCIA-MULTAS E JUROS                      0,00   
2            OUTRAS RECEITAS-PRIMARIAS-PRINCIPAL                      0,00   
3     TAXA DE CONTROLE E FISC.AMBIENTAL-MUL.JUR.                      0,00   
4             ONUS DE SUCUMBENCIA-MULTAS E JUROS                      0,00   

  VALOR LANÇADO VALOR REALIZADO PERCENTUAL REALIZADO DATA LANÇAMENTO  \
0          0,00       154948,98                 0,00      04/04/2016   
1          0,00           80,84                 0,00      08/12/2016   
2          0,00         4996,11                 0,00      22/08/2016   
3          0,00            0,77                 0,00      21/03/2016   
4          0,00           23,45                 0,00      04/11/2016   

   ANO EXERCÍCIO  
0           2016  
1           2016  
2           2016  
3           2016  
4           2016

In [141]:
#Me saco el numero de filas y columnas de este dataframe
df_2016.shape

(194533, 16)

Para 2016 contamos con 194533 filas y 16 columnas

In [142]:
#Para observar nulos
df_2016.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 194533 entries, 0 to 194532
Data columns (total 16 columns):
 #   Column                     Non-Null Count   Dtype 
---  ------                     --------------   ----- 
 0   CÓDIGO ÓRGÃO SUPERIOR      194533 non-null  int64 
 1   NOME ÓRGÃO SUPERIOR        194533 non-null  object
 2   CÓDIGO ÓRGÃO               194533 non-null  int64 
 3   NOME ÓRGÃO                 194533 non-null  object
 4   CÓDIGO UNIDADE GESTORA     194533 non-null  int64 
 5   NOME UNIDADE GESTORA       194533 non-null  object
 6   CATEGORIA ECONÔMICA        194533 non-null  object
 7   ORIGEM RECEITA             194533 non-null  object
 8   ESPÉCIE RECEITA            194533 non-null  object
 9   DETALHAMENTO               194533 non-null  object
 10  VALOR PREVISTO ATUALIZADO  194533 non-null  object
 11  VALOR LANÇADO              194533 non-null  object
 12  VALOR REALIZADO            194533 non-null  object
 13  PERCENTUAL REALIZADO       194533 non-null  

In [143]:
#Comprobar si hay duplicados por todas las columnas
df_2016.duplicated().sum()

np.int64(0)

Para 2016 contamos con nulos en la columna de la fecha de registro y no hay duplicados

In [146]:
#Sacar el porcentaje de nulos para la columna que los contiene
porcentaje_nulos = (df_2016["DATA LANÇAMENTO"].isnull().sum() / df_2016.shape[0])*100
porcentaje_nulos

np.float64(0.04215223124097197)

La columna cuenta con un 0,04% de nulos, lo que es prácticamente insignificante pero voy a observar en que filas están estos nulos

In [157]:
#Observación de los nulos
df_nulos_2016 = df_2016[df_2016["DATA LANÇAMENTO"].isnull()]
df_nulos_2016

CÓDIGO ÓRGÃO SUPERIOR                            NOME ÓRGÃO SUPERIOR  \
18141                   24000  Ministério da Ciência, Tecnologia, Inovações    
37397                   25000                         Ministério da Economia   
37458                   25000                         Ministério da Economia   
38173                   25000                         Ministério da Economia   
38403                   25000                         Ministério da Economia   
...                       ...                                            ...   
162740                  41000                    Ministério das Comunicações   
171364                  32000                  Ministério de Minas e Energia   
172316                  32000                  Ministério de Minas e Energia   
186873                  44000                    Ministério do Meio Ambiente   
189917                  54000                          Ministério do Turismo   

        CÓDIGO ÓRGÃO                                         NOME ÓRGÃO  \
18141          20501  Conselho Nacional de Desenvolvimento Científic...   
37397          25000  Ministério da Economia - Unidades com vínculo ...   
37458          25000  Ministério da Economia - Unidades com vínculo ...   
38173          25000  Ministério da Economia - Unidades com vínculo ...   
38403          25000  Ministério da Economia - Unidades com vínculo ...   
...              ...                                                ...   
162740         41000  Ministério das Comunicações - Unidades com vín...   
171364         29208         Companhia de Pesquisa de Recursos Minerais   
172316         32210               Agência Nacional de Energia Elétrica   
186873         20701  Instituto Brasileiro do Meio Ambiente e dos Re...   
189917         20203                         Agência Nacional do Cinema   

        CÓDIGO UNIDADE GESTORA                           NOME UNIDADE GESTORA  \
18141                   364102     CONSELHO NAC DE DESENV CIENT E TECNOLOGICO   
37397                   170010    SECRETARIA DA RECEITA FEDERAL DO BRASIL-RFB   
37458                   170010    SECRETARIA DA RECEITA FEDERAL DO BRASIL-RFB   
38173                   170010    SECRETARIA DA RECEITA FEDERAL DO BRASIL-RFB   
38403                   170010    SECRETARIA DA RECEITA FEDERAL DO BRASIL-RFB   
...                        ...                                            ...   
162740                  410003       COORDENACAO-GERAL DE RECURSOS LOGISTICOS   
171364                  495001     COMPANHIA DE PESQUISA DE RECURSOS MINERAIS   
172316                  323028           AGENCIA NACIONAL DE ENERGIA ELETRICA   
186873                  193107  IBAMA - SUPERINTENDENCIA DO ESPIRITO SANTO/ES   
189917                  203003                     AGENCIA NACIONAL DO CINEMA   

        CATEGORIA ECONÔMICA                               ORIGEM RECEITA  \
18141   Receitas de Capital                         Operações de Crédito   
37397    Receitas Correntes  Impostos, Taxas e Contribuições de Melhoria   
37458    Receitas Correntes  Impostos, Taxas e Contribuições de Melhoria   
38173    Receitas Correntes  Impostos, Taxas e Contribuições de Melhoria   
38403    Receitas Correntes  Impostos, Taxas e Contribuições de Melhoria   
...                     ...                                          ...   
162740  Receitas de Capital                         Operações de Crédito   
171364   Receitas Correntes           Receitas Correntes - a classificar   
172316   Receitas Correntes  Impostos, Taxas e Contribuições de Melhoria   
186873   Receitas Correntes  Impostos, Taxas e Contribuições de Melhoria   
189917   Receitas Correntes                                Contribuições   

                               ESPÉCIE RECEITA  \
18141   Operações de crédito - mercado interno   
37397                                 Impostos   
37458                                 Impostos   
38173                                 Impostos   
38403     

No observo ningún patrón en los nulos. Al ser una fecha que no es un dato sencillo de poder rellenar y al ser tan pocos los registros nulos, considero que podría mantenerlos tal cual están y trabajar con ellos ya que es un % muy mínimo. Borrar estas filas no tendría sentido porque estaríamos perdiendo información de otras columnas que si nos puede interesar.

Para 2017

In [148]:
#Leo el csv de datos de 2017
df_2017 = pd.read_csv("datos/datos-2017.csv", sep=";", encoding="ISO-8859-1")
df_2017.head()

CÓDIGO ÓRGÃO SUPERIOR       NOME ÓRGÃO SUPERIOR  CÓDIGO ÓRGÃO  \
0                  63000  Advocacia-Geral da União         63000   
1                  63000  Advocacia-Geral da União         63000   
2                  63000  Advocacia-Geral da União         63000   
3                  63000  Advocacia-Geral da União         63000   
4                  63000  Advocacia-Geral da União         63000   

                                          NOME ÓRGÃO  CÓDIGO UNIDADE GESTORA  \
0  Advocacia-Geral da União - Unidades com víncul...                  110060   
1  Advocacia-Geral da União - Unidades com víncul...                  110060   
2  Advocacia-Geral da União - Unidades com víncul...                  110060   
3  Advocacia-Geral da União - Unidades com víncul...                  110060   
4  Advocacia-Geral da União - Unidades com víncul...                  110060   

                            NOME UNIDADE GESTORA CATEGORIA ECONÔMICA  \
0  COORD. GERAL DE ORC. FIN. E ANAL. CONT. - AGU  Receitas Correntes   
1  COORD. GERAL DE ORC. FIN. E ANAL. CONT. - AGU  Receitas Correntes   
2  COORD. GERAL DE ORC. FIN. E ANAL. CONT. - AGU  Receitas Correntes   
3  COORD. GERAL DE ORC. FIN. E ANAL. CONT. - AGU  Receitas Correntes   
4  COORD. GERAL DE ORC. FIN. E ANAL. CONT. - AGU  Receitas Correntes   

              ORIGEM RECEITA            ESPÉCIE RECEITA  \
0  Outras Receitas Correntes  Demais receitas correntes   
1  Outras Receitas Correntes  Demais receitas correntes   
2  Outras Receitas Correntes  Demais receitas correntes   
3  Outras Receitas Correntes  Demais receitas correntes   
4  Outras Receitas Correntes  Demais receitas correntes   

                          DETALHAMENTO VALOR PREVISTO ATUALIZADO  \
0  OUTRAS RECEITAS-PRIMARIAS-PRINCIPAL                      0,00   
1  OUTRAS RECEITAS-PRIMARIAS-PRINCIPAL                      0,00   
2  OUTRAS RECEITAS-PRIMARIAS-PRINCIPAL                      0,00   
3  OUTRAS RECEITAS-PRIMARIAS-PRINCIPAL                      0,00   
4  OUTRAS RECEITAS-PRIMARIAS-PRINCIPAL                      0,00   

  VALOR LANÇADO VALOR REALIZADO PERCENTUAL REALIZADO DATA LANÇAMENTO  \
0          0,00          198,00                 0,00      29/08/2017   
1          0,00         2093,50                 0,00      10/02/2017   
2          0,00         9329,40                 0,00      27/11/2017   
3          0,00         2585,00                 0,00      13/11/2017   
4          0,00         3936,00                 0,00      03/05/2017   

   ANO EXERCÍCIO  
0           2017  
1           2017  
2           2017  
3           2017  
4           2017

In [150]:
#Me saco el numero de filas y columnas de este dataframe
df_2017.shape

(190479, 16)

Para 2017 contamos con 190479 filas y 16 columnas

In [151]:
#Para observar nulos
df_2017.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 190479 entries, 0 to 190478
Data columns (total 16 columns):
 #   Column                     Non-Null Count   Dtype 
---  ------                     --------------   ----- 
 0   CÓDIGO ÓRGÃO SUPERIOR      190479 non-null  int64 
 1   NOME ÓRGÃO SUPERIOR        190479 non-null  object
 2   CÓDIGO ÓRGÃO               190479 non-null  int64 
 3   NOME ÓRGÃO                 190479 non-null  object
 4   CÓDIGO UNIDADE GESTORA     190479 non-null  int64 
 5   NOME UNIDADE GESTORA       190479 non-null  object
 6   CATEGORIA ECONÔMICA        190479 non-null  object
 7   ORIGEM RECEITA             190479 non-null  object
 8   ESPÉCIE RECEITA            190479 non-null  object
 9   DETALHAMENTO               190479 non-null  object
 10  VALOR PREVISTO ATUALIZADO  190479 non-null  object
 11  VALOR LANÇADO              190479 non-null  object
 12  VALOR REALIZADO            190479 non-null  object
 13  PERCENTUAL REALIZADO       190479 non-null  

In [153]:
#Comprobar si hay duplicados por todas las columnas
df_2017.duplicated().sum()

np.int64(0)

Para 2017 ocurre algo parecido a 2016. No hay duplicados pero contamos con algunos nulos en la columna de fecha de registro

In [154]:
#Sacar el porcentaje de nulos para la columna que los contiene
porcentaje_nulos = (df_2017["DATA LANÇAMENTO"].isnull().sum() / df_2017.shape[0])*100
porcentaje_nulos

np.float64(0.045149334047322805)

In [158]:
#Observación de los nulos
df_nulos_2017 = df_2017[df_2017["DATA LANÇAMENTO"].isnull()]
df_nulos_2017

CÓDIGO ÓRGÃO SUPERIOR                            NOME ÓRGÃO SUPERIOR  \
17262                   24000  Ministério da Ciência, Tecnologia, Inovações    
36136                   25000                         Ministério da Economia   
36570                   25000                         Ministério da Economia   
36611                   25000                         Ministério da Economia   
37932                   25000                         Ministério da Economia   
...                       ...                                            ...   
160937                  41000                    Ministério das Comunicações   
170022                  32000                  Ministério de Minas e Energia   
170189                  32000                  Ministério de Minas e Energia   
183381                  44000                    Ministério do Meio Ambiente   
186669                  54000                          Ministério do Turismo   

        CÓDIGO ÓRGÃO                                         NOME ÓRGÃO  \
17262          20501  Conselho Nacional de Desenvolvimento Científic...   
36136          25000  Ministério da Economia - Unidades com vínculo ...   
36570          25000  Ministério da Economia - Unidades com vínculo ...   
36611          25000  Ministério da Economia - Unidades com vínculo ...   
37932          25000  Ministério da Economia - Unidades com vínculo ...   
...              ...                                                ...   
160937         41000  Ministério das Comunicações - Unidades com vín...   
170022         32210               Agência Nacional de Energia Elétrica   
170189         32210               Agência Nacional de Energia Elétrica   
183381         20701  Instituto Brasileiro do Meio Ambiente e dos Re...   
186669         20203                         Agência Nacional do Cinema   

        CÓDIGO UNIDADE GESTORA                           NOME UNIDADE GESTORA  \
17262                   364102     CONSELHO NAC DE DESENV CIENT E TECNOLOGICO   
36136                   170010    SECRETARIA DA RECEITA FEDERAL DO BRASIL-RFB   
36570                   170010    SECRETARIA DA RECEITA FEDERAL DO BRASIL-RFB   
36611                   170010    SECRETARIA DA RECEITA FEDERAL DO BRASIL-RFB   
37932                   170010    SECRETARIA DA RECEITA FEDERAL DO BRASIL-RFB   
...                        ...                                            ...   
160937                  410003       COORDENACAO-GERAL DE RECURSOS LOGISTICOS   
170022                  323028           AGENCIA NACIONAL DE ENERGIA ELETRICA   
170189                  323028           AGENCIA NACIONAL DE ENERGIA ELETRICA   
183381                  193107  IBAMA - SUPERINTENDENCIA DO ESPIRITO SANTO/ES   
186669                  203003                     AGENCIA NACIONAL DO CINEMA   

        CATEGORIA ECONÔMICA                               ORIGEM RECEITA  \
17262   Receitas de Capital                         Operações de Crédito   
36136    Receitas Correntes                                Contribuições   
36570    Receitas Correntes                                Contribuições   
36611    Receitas Correntes                                Contribuições   
37932    Receitas Correntes  Impostos, Taxas e Contribuições de Melhoria   
...                     ...                                          ...   
160937  Receitas de Capital                         Operações de Crédito   
170022   Receitas Correntes  Impostos, Taxas e Contribuições de Melhoria   
170189   Receitas Correntes  Impostos, Taxas e Contribuições de Melhoria   
183381   Receitas Correntes  Impostos, Taxas e Contribuições de Melhoria   
186669   Receitas Correntes                                Contribuições   

                               ESPÉCIE RECEITA  \
17262   Operações de crédito - mercado interno   
36136                    Contribuições sociais   
36570                    Contribuições sociais   
36611                    Contribuições sociais   
37932     

En este caso pasa parecido y voy a optar por tomar la misma decisión que es trabajar con ellos ya que el % es mínimo

Para 2018

In [167]:
#Leo el csv de datos de 2018
df_2018 = pd.read_csv("datos/datos-2018.csv", sep=";", encoding="ISO-8859-1")
df_2018.head()

CÓDIGO ÓRGÃO SUPERIOR       NOME ÓRGÃO SUPERIOR  CÓDIGO ÓRGÃO  \
0                  63000  Advocacia-Geral da União         63000   
1                  63000  Advocacia-Geral da União         63000   
2                  63000  Advocacia-Geral da União         63000   
3                  63000  Advocacia-Geral da União         63000   
4                  63000  Advocacia-Geral da União         63000   

                                          NOME ÓRGÃO  CÓDIGO UNIDADE GESTORA  \
0  Advocacia-Geral da União - Unidades com víncul...                  110060   
1  Advocacia-Geral da União - Unidades com víncul...                  110060   
2  Advocacia-Geral da União - Unidades com víncul...                  110060   
3  Advocacia-Geral da União - Unidades com víncul...                  110060   
4  Advocacia-Geral da União - Unidades com víncul...                  110060   

                            NOME UNIDADE GESTORA CATEGORIA ECONÔMICA  \
0  COORD. GERAL DE ORC. FIN. E ANAL. CONT. - AGU  Receitas Correntes   
1  COORD. GERAL DE ORC. FIN. E ANAL. CONT. - AGU  Receitas Correntes   
2  COORD. GERAL DE ORC. FIN. E ANAL. CONT. - AGU  Receitas Correntes   
3  COORD. GERAL DE ORC. FIN. E ANAL. CONT. - AGU  Receitas Correntes   
4  COORD. GERAL DE ORC. FIN. E ANAL. CONT. - AGU  Receitas Correntes   

                       ORIGEM RECEITA  \
0  Receitas Correntes - a classificar   
1            Transferências Correntes   
2  Receitas Correntes - a classificar   
3  Receitas Correntes - a classificar   
4  Receitas Correntes - a classificar   

                                 ESPÉCIE RECEITA  \
0             Receitas Correntes - a classificar   
1  Transferências dos Estados e do Distrito Fede   
2             Receitas Correntes - a classificar   
3             Receitas Correntes - a classificar   
4             Receitas Correntes - a classificar   

                                    DETALHAMENTO VALOR PREVISTO ATUALIZADO  \
0             Receitas Correntes - a classificar                      0,00   
1  TRANSF.DOS ESTADOS,DF E SUAS ENTIDADES-PRINC.                      0,00   
2             Receitas Correntes - a classificar                      0,00   
3             Receitas Correntes - a classificar                      0,00   
4             Receitas Correntes - a classificar                      0,00   

  VALOR LANÇADO VALOR REALIZADO PERCENTUAL REALIZADO DATA LANÇAMENTO  \
0          0,00        -1693,72                 0,00      25/07/2018   
1          0,00         2612,50                 0,00      02/08/2018   
2          0,00         -258,00                 0,00      27/02/2018   
3          0,00          258,00                 0,00      26/02/2018   
4          0,00         4747,33                 0,00      05/03/2018   

   ANO EXERCÍCIO  
0           2018  
1           2018  
2           2018  
3           2018  
4           2018

In [168]:
#Me saco el numero de filas y columnas de este dataframe
df_2018.shape

(173944, 16)

Para 2018 contamos con 173944 filas y 16 columnas

In [169]:
#Para observar nulos
df_2018.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 173944 entries, 0 to 173943
Data columns (total 16 columns):
 #   Column                     Non-Null Count   Dtype 
---  ------                     --------------   ----- 
 0   CÓDIGO ÓRGÃO SUPERIOR      173944 non-null  int64 
 1   NOME ÓRGÃO SUPERIOR        173944 non-null  object
 2   CÓDIGO ÓRGÃO               173944 non-null  int64 
 3   NOME ÓRGÃO                 173944 non-null  object
 4   CÓDIGO UNIDADE GESTORA     173944 non-null  int64 
 5   NOME UNIDADE GESTORA       173944 non-null  object
 6   CATEGORIA ECONÔMICA        173944 non-null  object
 7   ORIGEM RECEITA             173944 non-null  object
 8   ESPÉCIE RECEITA            173944 non-null  object
 9   DETALHAMENTO               173944 non-null  object
 10  VALOR PREVISTO ATUALIZADO  173944 non-null  object
 11  VALOR LANÇADO              173944 non-null  object
 12  VALOR REALIZADO            173944 non-null  object
 13  PERCENTUAL REALIZADO       173944 non-null  

In [170]:
#Comprobar si hay duplicados por todas las columnas
df_2018.duplicated().sum()

np.int64(0)

De nuevo, para 2018 tampoco hay duplicados pero contamos con nulos en la misma columna, la de fecha de registro

In [171]:
#Sacar el porcentaje de nulos para la columna que los contiene
porcentaje_nulos = (df_2018["DATA LANÇAMENTO"].isnull().sum() / df_2018.shape[0])*100
porcentaje_nulos

np.float64(0.06611323184473164)

In [172]:
#Observación de los nulos
df_nulos_2018 = df_2018[df_2018["DATA LANÇAMENTO"].isnull()]
df_nulos_2018

CÓDIGO ÓRGÃO SUPERIOR                            NOME ÓRGÃO SUPERIOR  \
8310                    24000  Ministério da Ciência, Tecnologia, Inovações    
29204                   25000                         Ministério da Economia   
29297                   25000                         Ministério da Economia   
29666                   25000                         Ministério da Economia   
29723                   25000                         Ministério da Economia   
...                       ...                                            ...   
157405                  32000                  Ministério de Minas e Energia   
164643                  44000                    Ministério do Meio Ambiente   
164646                  44000                    Ministério do Meio Ambiente   
168923                  44000                    Ministério do Meio Ambiente   
171452                  54000                          Ministério do Turismo   

        CÓDIGO ÓRGÃO                                         NOME ÓRGÃO  \
8310           24000  Ministério da Ciência, Tecnologia, Inovações e...   
29204          25000  Ministério da Economia - Unidades com vínculo ...   
29297          25000  Ministério da Economia - Unidades com vínculo ...   
29666          25000  Ministério da Economia - Unidades com vínculo ...   
29723          25000  Ministério da Economia - Unidades com vínculo ...   
...              ...                                                ...   
157405         32210               Agência Nacional de Energia Elétrica   
164643         20701  Instituto Brasileiro do Meio Ambiente e dos Re...   
164646         20701  Instituto Brasileiro do Meio Ambiente e dos Re...   
168923         20701  Instituto Brasileiro do Meio Ambiente e dos Re...   
171452         20203                         Agência Nacional do Cinema   

        CÓDIGO UNIDADE GESTORA                           NOME UNIDADE GESTORA  \
8310                    240101       COORDENACAO-GERAL DE RECURSOS LOGISTICOS   
29204                   170010    SECRETARIA DA RECEITA FEDERAL DO BRASIL-RFB   
29297                   170010    SECRETARIA DA RECEITA FEDERAL DO BRASIL-RFB   
29666                   170010    SECRETARIA DA RECEITA FEDERAL DO BRASIL-RFB   
29723                   170010    SECRETARIA DA RECEITA FEDERAL DO BRASIL-RFB   
...                        ...                                            ...   
157405                  323028           AGENCIA NACIONAL DE ENERGIA ELETRICA   
164643                  193129       IBAMA - SUPERINTENDENCIA DE SAO PAULO/SP   
164646                  193120   IBAMA - SUPERINTENDENCIA RIO GRANDE NORTE/RN   
168923                  193107  IBAMA - SUPERINTENDENCIA DO ESPIRITO SANTO/ES   
171452                  203003                     AGENCIA NACIONAL DO CINEMA   

        CATEGORIA ECONÔMICA                               ORIGEM RECEITA  \
8310    Receitas de Capital                         Operações de Crédito   
29204    Receitas Correntes                                Contribuições   
29297    Receitas Correntes                                Contribuições   
29666    Receitas Correntes                                Contribuições   
29723    Receitas Correntes  Impostos, Taxas e Contribuições de Melhoria   
...                     ...                                          ...   
157405   Receitas Correntes                    Outras Receitas Correntes   
164643   Receitas Correntes  Impostos, Taxas e Contribuições de Melhoria   
164646   Receitas Correntes  Impostos, Taxas e Contribuições de Melhoria   
168923   Receitas Correntes  Impostos, Taxas e Contribuições de Melhoria   
171452   Receitas Correntes                                Contribuições   

                                      ESPÉCIE RECEITA  \
8310           Operações de crédito - mercado interno   
29204                           Contribuições sociais   
29297                        Contribuições econômicas   
29666                        Con

De nuevo, opto por tomar la misma decisión con los nulos, mantenerlos

Para 2019

In [176]:
#Leo el csv de datos de 2019
df_2019 = pd.read_csv("datos/datos-2019.csv", sep=";", encoding="ISO-8859-1")
df_2019.head()

CÓDIGO ÓRGÃO SUPERIOR       NOME ÓRGÃO SUPERIOR  CÓDIGO ÓRGÃO  \
0                  63000  Advocacia-Geral da União         63000   
1                  63000  Advocacia-Geral da União         63000   
2                  63000  Advocacia-Geral da União         63000   
3                  63000  Advocacia-Geral da União         63000   
4                  63000  Advocacia-Geral da União         63000   

                                          NOME ÓRGÃO  CÓDIGO UNIDADE GESTORA  \
0  Advocacia-Geral da União - Unidades com víncul...                  110060   
1  Advocacia-Geral da União - Unidades com víncul...                  110060   
2  Advocacia-Geral da União - Unidades com víncul...                  110060   
3  Advocacia-Geral da União - Unidades com víncul...                  110060   
4  Advocacia-Geral da União - Unidades com víncul...                  110060   

                            NOME UNIDADE GESTORA CATEGORIA ECONÔMICA  \
0  COORD. GERAL DE ORC. FIN. E ANAL. CONT. - AGU  Receitas Correntes   
1  COORD. GERAL DE ORC. FIN. E ANAL. CONT. - AGU  Receitas Correntes   
2  COORD. GERAL DE ORC. FIN. E ANAL. CONT. - AGU  Receitas Correntes   
3  COORD. GERAL DE ORC. FIN. E ANAL. CONT. - AGU  Receitas Correntes   
4  COORD. GERAL DE ORC. FIN. E ANAL. CONT. - AGU  Receitas Correntes   

        ORIGEM RECEITA                               ESPÉCIE RECEITA  \
0  Receita de Serviços  Serviços Administrativos e Comerciais Gerais   
1  Receita de Serviços  Serviços Administrativos e Comerciais Gerais   
2  Receita de Serviços  Serviços Administrativos e Comerciais Gerais   
3  Receita de Serviços  Serviços Administrativos e Comerciais Gerais   
4  Receita de Serviços  Serviços Administrativos e Comerciais Gerais   

                                    DETALHAMENTO VALOR PREVISTO ATUALIZADO  \
0  INSCR.EM CONCURSOS E PROC.SELETIVOS-PRINCIPAL                      0,00   
1  INSCR.EM CONCURSOS E PROC.SELETIVOS-PRINCIPAL                      0,00   
2  INSCR.EM CONCURSOS E PROC.SELETIVOS-PRINCIPAL                      0,00   
3  INSCR.EM CONCURSOS E PROC.SELETIVOS-PRINCIPAL                      0,00   
4  INSCR.EM CONCURSOS E PROC.SELETIVOS-PRINCIPAL                      0,00   

  VALOR LANÇADO VALOR REALIZADO PERCENTUAL REALIZADO DATA LANÇAMENTO  \
0          0,00          -95,00                 0,00      12/06/2019   
1          0,00         -380,00                 0,00      07/05/2019   
2          0,00          380,00                 0,00      01/03/2019   
3          0,00       -13490,00                 0,00      27/02/2019   
4          0,00         -855,00                 0,00      18/04/2019   

   ANO EXERCÍCIO  
0           2019  
1           2019  
2           2019  
3           2019  
4           2019

In [178]:
#Me saco el numero de filas y columnas de este dataframe
df_2019.shape

(176828, 16)

Para 2019 cuento con 176828 filas y 16 columnas

In [179]:
#Para observar nulos
df_2019.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 176828 entries, 0 to 176827
Data columns (total 16 columns):
 #   Column                     Non-Null Count   Dtype 
---  ------                     --------------   ----- 
 0   CÓDIGO ÓRGÃO SUPERIOR      176828 non-null  int64 
 1   NOME ÓRGÃO SUPERIOR        176828 non-null  object
 2   CÓDIGO ÓRGÃO               176828 non-null  int64 
 3   NOME ÓRGÃO                 176828 non-null  object
 4   CÓDIGO UNIDADE GESTORA     176828 non-null  int64 
 5   NOME UNIDADE GESTORA       176828 non-null  object
 6   CATEGORIA ECONÔMICA        176828 non-null  object
 7   ORIGEM RECEITA             176828 non-null  object
 8   ESPÉCIE RECEITA            176828 non-null  object
 9   DETALHAMENTO               176828 non-null  object
 10  VALOR PREVISTO ATUALIZADO  176828 non-null  object
 11  VALOR LANÇADO              176828 non-null  object
 12  VALOR REALIZADO            176828 non-null  object
 13  PERCENTUAL REALIZADO       176828 non-null  

In [180]:
#Comprobar si hay duplicados por todas las columnas
df_2019.duplicated().sum()

np.int64(0)

Para 2019 de nuevo no hay duplicados pero hay valores nulos en la columna de fecha de registro

In [181]:
#Sacar el porcentaje de nulos para la columna que los contiene
porcentaje_nulos = (df_2019["DATA LANÇAMENTO"].isnull().sum() / df_2019.shape[0])*100
porcentaje_nulos

np.float64(0.048634831587757596)

El % de nulos sigue siendo mínimo por lo que de nuevo los voy a mantener como están. Pero al observar que está ocurriendo algo parecido en todos los años puede ser que haya una época en todos los años en los que por alguna razón la fecha de registro se está perdiendo y sería algo a tomar en cuenta. Voy a intentar observar esto.

In [115]:
#Leo el csv de datos de 2020
df_2020 = pd.read_csv("datos/datos-2020.csv", sep=";", index_col=0, encoding="ISO-8859-1")
df_2020.head()

NOME ÓRGÃO SUPERIOR  CÓDIGO ÓRGÃO  \
CÓDIGO ÓRGÃO SUPERIOR                                           
63000                  Advocacia-Geral da União         63000   
63000                  Advocacia-Geral da União         63000   
63000                  Advocacia-Geral da União         63000   
63000                  Advocacia-Geral da União         63000   
63000                  Advocacia-Geral da União         63000   

                                                              NOME ÓRGÃO  \
CÓDIGO ÓRGÃO SUPERIOR                                                      
63000                  Advocacia-Geral da União - Unidades com víncul...   
63000                  Advocacia-Geral da União - Unidades com víncul...   
63000                  Advocacia-Geral da União - Unidades com víncul...   
63000                  Advocacia-Geral da União - Unidades com víncul...   
63000                  Advocacia-Geral da União - Unidades com víncul...   

                       CÓDIGO UNIDADE GESTORA  \
CÓDIGO ÓRGÃO SUPERIOR                           
63000                                  110060   
63000                                  110060   
63000                                  110060   
63000                                  110060   
63000                                  110060   

                                                NOME UNIDADE GESTORA  \
CÓDIGO ÓRGÃO SUPERIOR                                                  
63000                  COORD. GERAL DE ORC. FIN. E ANAL. CONT. - AGU   
63000                  COORD. GERAL DE ORC. FIN. E ANAL. CONT. - AGU   
63000                  COORD. GERAL DE ORC. FIN. E ANAL. CONT. - AGU   
63000                  COORD. GERAL DE ORC. FIN. E ANAL. CONT. - AGU   
63000                  COORD. GERAL DE ORC. FIN. E ANAL. CONT. - AGU   

                      CATEGORIA ECONÔMICA             ORIGEM RECEITA  \
CÓDIGO ÓRGÃO SUPERIOR                                                  
63000                  Receitas Correntes  Outras Receitas Correntes   
63000                  Receitas Correntes  Outras Receitas Correntes   
63000                  Receitas Correntes  Outras Receitas Correntes   
63000                  Receitas Correntes  Outras Receitas Correntes   
63000                  Receitas Correntes  Outras Receitas Correntes   

                                                   ESPÉCIE RECEITA  \
CÓDIGO ÓRGÃO SUPERIOR                                                
63000                  Indenizações, restituições e ressarcimentos   
63000                  Indenizações, restituições e ressarcimentos   
63000                  Indenizações, restituições e ressarcimentos   
63000                  Indenizações, restituições e ressarcimentos   
63000                  Indenizações, restituições e ressarcimentos   

                                                       DETALHAMENTO  \
CÓDIGO ÓRGÃO SUPERIOR                                                 
63000                  RESTITUIÇÃO DE CONVÊNIOS-PRIMÁRIAS-DÍV.ATIVA   
63000                  RESTITUIÇÃO DE CONVÊNIOS-PRIMÁRIAS-DÍV.ATIVA   
63000                  RESTITUIÇÃO DE CONVÊNIOS-PRIMÁRIAS-DÍV.ATIVA   
63000                  RESTITUIÇÃO DE CONVÊNIOS-PRIMÁRIAS-DÍV.ATIVA   
63000                  RESTITUIÇÃO DE CONVÊNIOS-PRIMÁRIAS-DÍV.ATIVA   

                      VALOR PREVISTO ATUALIZADO VALOR LANÇADO VALOR REALIZADO  \
CÓDIGO ÓRGÃO SUPERIOR                                                           
63000                                      0,00          0,00         5516,90   
63000                                      0,00          0,00         2946,81   
63000                                      0,00          0,00         1541,53   
63000                                      0,00          0,00            0,00   
63000                                      0,00          0,00         4701,23   

                      PERCENTUAL REALIZADO DATA LANÇAMENTO  ANO EXERCÍCIO  
CÓDIGO ÓRGÃO SUPERIOR                               

In [116]:
#Leo el csv de datos de 2021
df_2021 = pd.read_csv("datos/datos-2021.csv", sep=";", index_col=0, encoding="ISO-8859-1")
df_2021.head()

NOME ÓRGÃO SUPERIOR  \
CÓDIGO ÓRGÃO SUPERIOR                                                  
63000                                       Advocacia-Geral da União   
63000                                       Advocacia-Geral da União   
37000                                   Controladoria-Geral da União   
22000                  Ministério da Agricultura, Pecuária e Abastec   
22000                  Ministério da Agricultura, Pecuária e Abastec   

                       CÓDIGO ÓRGÃO  \
CÓDIGO ÓRGÃO SUPERIOR                 
63000                         63000   
63000                         63000   
37000                         37000   
22000                         22211   
22000                         22211   

                                                              NOME ÓRGÃO  \
CÓDIGO ÓRGÃO SUPERIOR                                                      
63000                  Advocacia-Geral da União - Unidades com víncul...   
63000                  Advocacia-Geral da União - Unidades com víncul...   
37000                  Controladoria-Geral da União - Unidades com ví...   
22000                                Companhia Nacional de Abastecimento   
22000                                Companhia Nacional de Abastecimento   

                       CÓDIGO UNIDADE GESTORA  \
CÓDIGO ÓRGÃO SUPERIOR                           
63000                                  110060   
63000                                  110060   
37000                                  370001   
22000                                  135100   
22000                                  135100   

                                                NOME UNIDADE GESTORA  \
CÓDIGO ÓRGÃO SUPERIOR                                                  
63000                  COORD. GERAL DE ORC. FIN. E ANAL. CONT. - AGU   
63000                  COORD. GERAL DE ORC. FIN. E ANAL. CONT. - AGU   
37000                               DIRETORIA DE GESTAO INTERNA(CGU)   
22000                            COMPANHIA NACIONAL DE ABASTECIMENTO   
22000                            COMPANHIA NACIONAL DE ABASTECIMENTO   

                       CATEGORIA ECONÔMICA  \
CÓDIGO ÓRGÃO SUPERIOR                        
63000                  Receitas de Capital   
63000                   Receitas Correntes   
37000                  Receitas de Capital   
22000                   Receitas Correntes   
22000                   Receitas Correntes   

                                           ORIGEM RECEITA  \
CÓDIGO ÓRGÃO SUPERIOR                                       
63000                                Operações de Crédito   
63000                  Receitas Correntes - a classificar   
37000                                Operações de Crédito   
22000                                 Receita de Serviços   
22000                                 Receita de Serviços   

                                                    ESPÉCIE RECEITA  \
CÓDIGO ÓRGÃO SUPERIOR                                                 
63000                        Operações de crédito - mercado interno   
63000                            Receitas Correntes - a classificar   
37000                        Operações de crédito - mercado interno   
22000                  Serviços Administrativos e Comerciais Gerais   
22000                  Serviços Administrativos e Comerciais Gerais   

                                                        DETALHAMENTO  \
CÓDIGO ÓRGÃO SUPERIOR                                                  
63000                   TITULOS DE RESPONS.TES.NAC.-MERC.INT.-PRINC.   
63000                             Receitas Correntes - a classificar   
37000                   TITULOS DE RESPONS.TES.NAC.-MERC.INT.-PRINC.   
22000                  SERV.ADMINISTRAT.E COMERCIAIS GERAIS-MUL.JUR.   
22000                  SERV.ADMINISTRAT.E COMERCIAIS GERAIS-MUL.JUR.   

                      VALOR PREVISTO ATUALIZADO VALOR LANÇADO VALOR REALIZADO  \
CÓDIGO ÓRGÃO SUPERIOR                                